tests/gocase/unit/pubsub/pubsub_test.go

```diff
@@ -0,0 +1,356 @@
+
+	ctx := context.Background()
+	rdb := srv.NewClient()
+	defer func() { require.NoError(t, rdb.Close()) }()
+
+	t.Run("PUBLISH/SUBSCRIBE PING", func(t *testing.T) {
+		pubsub := rdb.Subscribe(ctx, "somechannel")
+		receiveType(t, pubsub, &redis.Subscription{})
+		require.NoError(t, pubsub.Ping(ctx))
+		require.NoError(t, pubsub.Ping(ctx))
+		require.NoError(t, pubsub.Unsubscribe(ctx, "somechannel"))
+		require.Equal(t, "PONG", rdb.Ping(ctx).Val())
+		receiveType(t, pubsub, &redis.Pong{})
+		receiveType(t, pubsub, &redis.Pong{})
+	})
+
+	t.Run("PUBLISH/SUBSCRIBE basics", func(t *testing.T) {
+		// subscribe to two channels
+		pubsub := rdb.Subscribe(ctx, "chan1", "chan2")
+		require.EqualValues(t, 1, receiveType(t, pubsub, &redis.Subscription{}).Count)
+		require.EqualValues(t, 2, receiveType(t, pubsub, &redis.Subscription{}).Count)
+
+		require.EqualValues(t, 1, rdb.Publish(ctx, "chan1", "hello").Val())
+		require.EqualValues(t, 1, rdb.Publish(ctx, "chan2", "world").Val())
+
+		msg := receiveType(t, pubsub, &redis.Message{})
+		require.Equal(t, "chan1", msg.Channel)
+		require.Equal(t, "hello", msg.Payload)
+		msg = receiveType(t, pubsub, &redis.Message{})
+		require.Equal(t, "chan2", msg.Channel)
+		require.Equal(t, "world", msg.Payload)
+
+		// unsubscribe from one of the channels
+		require.NoError(t, pubsub.Unsubscribe(ctx, "chan1"))
+		require.EqualValues(t, 0, rdb.Publish(ctx, "chan1", "hello").Val())
+		require.EqualValues(t, 1, rdb.Publish(ctx, "chan2", "world").Val())
+
+		// unsubscribe from the remaining channel
+		require.NoError(t, pubsub.Unsubscribe(ctx, "chan2"))
+		require.EqualValues(t, 0, rdb.Publish(ctx, "chan1", "hello").Val())
+		require.EqualValues(t, 0, rdb.Publish(ctx, "chan2", "world").Val())
+	})
+
+	t.Run("PUBLISH/SUBSCRIBE with two clients", func(t *testing.T) {
+		c1 := srv.NewClient()
+		defer func() { require.NoError(t, c1.Close()) }()
+		c2 := srv.NewClient()
+		defer func() { require.NoError(t, c2.Close()) }()
+
+		p1 := c1.Subscribe(ctx, "chan1")
+		p2 := c2.Subscribe(ctx, "chan1")
+		require.EqualValues(t, 1, receiveType(t, p1, &redis.Subscription{}).Count)
+		require.EqualValues(t, 1, receiveType(t, p2, &redis.Subscription{}).Count)
+
+		require.EqualValues(t, 2, rdb.Publish(ctx, "chan1", "hello").Val())
+		msg := receiveType(t, p1, &redis.Message{})
+		require.Equal(t, "chan1", msg.Channel)
+		require.Equal(t, "hello", msg.Payload)
+		msg = receiveType(t, p2, &redis.Message{})
+		require.Equal(t, "chan1", msg.Channel)
+		require.Equal(t, "hello", msg.Payload)
+	})
+
+	t.Run("PUBLISH/SUBSCRIBE after UNSUBSCRIBE without arguments", func(t *testing.T) {
+		pubsub := rdb.Subscribe(ctx, "chan1", "chan2", "chan3")
+		require.EqualValues(t, 1, receiveType(t, pubsub, &redis.Subscription{}).Count)
+		require.EqualValues(t, 2, receiveType(t, pubsub, &redis.Subscription{}).Count)
+		require.EqualValues(t, 3, receiveType(t, pubsub, &redis.Subscription{}).Count)
+
+		require.NoError(t, pubsub.Unsubscribe(ctx))
```

Comment:
@tisonkun looks like you need to add channels here - "chan1", "chan2", "chan3". Not sure why the test is not failing earlier though and you still receive `unsubscribe` events... Perhaps some magic by kvrocks? I wonder if Redis behaves the same.

Reply:
>from go-redis method docs:

Indeed! I honestly can't find where that is implemented and can't check it right away so don't trust it blindly :)

Perhaps it indeed unsubscribes thanks to Redis/kvrocks behavior, but does not reset the internal state of `PubSub` which can be a cause for surprises.

Reply:
Interesting. I can reproduce it locally and when I rename the channel name from "chan2" to "xx" the error vanished.

```diff
diff --git a/tests/gocase/unit/pubsub/pubsub_test.go b/tests/gocase/unit/pubsub/pubsub_test.go
index b48b10a..97b807b 100644
--- a/tests/gocase/unit/pubsub/pubsub_test.go
+++ b/tests/gocase/unit/pubsub/pubsub_test.go
@@ -73,11 +73,26 @@ func TestPubSub(t *testing.T) {
 
 		// unsubscribe from one of the channels
 		require.NoError(t, pubsub.Unsubscribe(ctx, "chan1"))
+		require.EqualValues(t, &redis.Subscription{
+			Kind:    "unsubscribe",
+			Channel: "chan1",
+			Count:   1,
+		}, receiveType(t, pubsub, &redis.Subscription{}))
 		require.EqualValues(t, 0, rdb.Publish(ctx, "chan1", "hello").Val())
 		require.EqualValues(t, 1, rdb.Publish(ctx, "chan2", "world").Val())
 
+		// drain incoming messages
+		// TODO: we should receive once here, but this bug cause the below unstable test steady to fail.
+		receiveType(t, pubsub, &redis.Message{})
+		receiveType(t, pubsub, &redis.Message{})
+
 		// unsubscribe from the remaining channel
 		require.NoError(t, pubsub.Unsubscribe(ctx, "chan2"))
+		require.EqualValues(t, &redis.Subscription{
+			Kind:    "unsubscribe",
+			Channel: "chan2",
+			Count:   0,
+		}, receiveType(t, pubsub, &redis.Subscription{}))
 		require.EqualValues(t, 0, rdb.Publish(ctx, "chan1", "hello").Val())
 		require.EqualValues(t, 0, rdb.Publish(ctx, "chan2", "world").Val())
 	})
@@ -103,7 +118,7 @@ func TestPubSub(t *testing.T) {
 	})
 
 	t.Run("PUBLISH/SUBSCRIBE after UNSUBSCRIBE without arguments", func(t *testing.T) {
-		pubsub := rdb.Subscribe(ctx, "chan1", "chan2", "chan3")
+		pubsub := rdb.Subscribe(ctx, "chan1", "xx", "chan3")
 		require.EqualValues(t, 1, receiveType(t, pubsub, &redis.Subscription{}).Count)
 		require.EqualValues(t, 2, receiveType(t, pubsub, &redis.Subscription{}).Count)
 		require.EqualValues(t, 3, receiveType(t, pubsub, &redis.Subscription{}).Count)
@@ -116,7 +131,7 @@ func TestPubSub(t *testing.T) {
 		}, receiveType(t, pubsub, &redis.Subscription{}))
 		require.EqualValues(t, &redis.Subscription{
 			Kind:    "unsubscribe",
-			Channel: "chan2",
+			Channel: "xx",
 			Count:   1,
 		}, receiveType(t, pubsub, &redis.Subscription{}))
 		require.EqualValues(t, &redis.Subscription{
@@ -126,7 +141,7 @@ func TestPubSub(t *testing.T) {
 		}, receiveType(t, pubsub, &redis.Subscription{}))
 
 		require.EqualValues(t, 0, rdb.Publish(ctx, "chan1", "hello").Val())
-		require.EqualValues(t, 0, rdb.Publish(ctx, "chan2", "hello").Val())
+		require.EqualValues(t, 0, rdb.Publish(ctx, "xx", "hello").Val())
 		require.EqualValues(t, 0, rdb.Publish(ctx, "chan3", "hello").Val())
 	})
```

Reply:
Forgive my comment above, it's about stateful tests.

Reply:
You can add `defer pubsub.Close()` to the tests to properly cleanup pubsub and go-redis should fix https://github.com/go-redis/redis/issues/2248

Reply:
So now I get the point:

Because we don't wait for unsubscription in L76, the "chan2" subscription will never be unsubscribed and leak.

Reply:
We may try to fail fast or reduce share state...But pubsub state are on the server side so the only approach is start a new server for each case?

Reply:
@vmihailenco thank you! Let me think of it.